worker/modelcache/worker.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelcache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/kr/pretty"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+
+	"github.com/juju/juju/core/cache"
+	"github.com/juju/juju/core/life"
+	"github.com/juju/juju/core/status"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/multiwatcher"
+)
+
+// Config describes the necessary fields for NewWorker.
+type Config struct {
+	Logger               Logger
+	StatePool            *state.StatePool
+	PrometheusRegisterer prometheus.Registerer
+	Cleanup              func()
+	// Notify is used primarily for testing, and is passed through
+	// to the cache.Controller. It is called every time the controller
+	// processes an event.
+	Notify func(interface{})
+}
+
+type cacheWorker struct {
+	config     Config
+	catacomb   catacomb.Catacomb
+	mu         sync.Mutex
+	controller *cache.Controller
+	changes    chan interface{}
+}
+
+// NewWorker creates a new cacheWorker, and starts an
+// all model watcher.
+func NewWorker(config Config) (worker.Worker, error) {
+	w := &cacheWorker{
+		config:  config,
+		changes: make(chan interface{}),
+	}
+	// In order to ensure that the cache.Controller has been created
+	// before the NewWorker function returns, we create a channel that the
+	// loop method closes once the controller has been created.
+	started := make(chan struct{})
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			return w.loop(started)
+		},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	select {
+	case <-started:
+	case <-time.After(5 * time.Second):
+		// The loop should have started, so if it hasn't, it is likely
+		// that it hit an error.
+		err := w.catacomb.Err()
+		if err != nil {
+			return nil, err
+		}
+		// If it hasn't stopped already, we're going to kill it.
+		err = errors.New("start took too long")
+		w.catacomb.Kill(err)
+		return nil, err
+	}
+	return w, nil
+}
+
+func (c *cacheWorker) getController() *cache.Controller {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.controller
+}
+
+func (c *cacheWorker) loop(started chan struct{}) error {
+	defer c.config.Cleanup()
+	pool := c.config.StatePool
+	allWatcher := pool.SystemState().WatchAllModels(pool)
+	defer allWatcher.Stop()
+
+	controller := cache.NewController(c.changes, c.config.Notify)
+	if err := c.catacomb.Add(controller); err != nil {
+		return errors.Trace(err)
+	}
+	c.mu.Lock()
+	c.controller = controller
+	c.mu.Unlock()
+
+	// We have now set the controller, so started enough.
+	close(started)
+
+	collector := cache.NewMetricsCollector(controller)
+	c.config.PrometheusRegisterer.Register(collector)
```

Comment:
Unhandled error.

Reply:
I don't think we really care about registation failures. Perhaps we should log.